Crash in SearchImageFragment

**Summary:** 
The Observable in SearchImageFragment has not been disposed as per the fragment lifecycle and because of which the app crashes (along with memory leaks), when RxCallbacks are received after the fragment is stopped/detached

**System logs:**

```
USER_COMMENT=French locale

ANDROID_VERSION=7.1.2
APP_VERSION_NAME=2.10.1.260
BRAND=google
PHONE_MODEL=Nexus 5
CUSTOM_DATA=
STACK_TRACE=io.reactivex.exceptions.OnErrorNotImplementedException: Attempt to invoke virtual method 'void android.view.View.setVisibility(int)' on a null object reference
at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:67)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)

at android.os.Handler.handleCallback(Handler.java:751)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:154)
at android.app.ActivityThread.main(ActivityThread.java:6186)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.setVisibility(int)' on a null object reference
at fr.free.nrw.commons.explore.images.SearchImageFragment.handleNoInternet(SearchImageFragment.java:233)
at fr.free.nrw.commons.explore.images.SearchImageFragment.updateImageList(SearchImageFragment.java:136)
at fr.free.nrw.commons.explore.SearchActivity.lambda$setTabs$1$SearchActivity(SearchActivity.java:112)
at fr.free.nrw.commons.explore.SearchActivity$$Lambda$1.accept(Unknown Source)
at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
... 10 more
java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.setVisibility(int)' on a null object reference
at fr.free.nrw.commons.explore.images.SearchImageFragment.handleNoInternet(SearchImageFragment.java:233)
at fr.free.nrw.commons.explore.images.SearchImageFragment.updateImageList(SearchImageFragment.java:136)
at fr.free.nrw.commons.explore.SearchActivity.lambda$setTabs$1$SearchActivity(SearchActivity.java:112)
at fr.free.nrw.commons.explore.SearchActivity$$Lambda$1.accept(Unknown Source)
at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)

at android.os.Handler.handleCallback(Handler.java:751)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:154)
at android.app.ActivityThread.main(ActivityThread.java:6186)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)  ```

**Device and Android version:** 
APP_VERSION_NAME=2.10.1.260
BRAND=google
PHONE_MODEL=Nexus 5

 
**Commons app version:** 
2.10.1.260

**Would you like to work on the issue?**
Yes

--- a/app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java
@@ -23,6 +23,7 @@ import fr.free.nrw.commons.explore.images.SearchImageFragment;
 import fr.free.nrw.commons.explore.recentsearches.RecentSearchesFragment;
 import fr.free.nrw.commons.media.MediaDetailPagerFragment;
 import fr.free.nrw.commons.theme.NavigationBaseActivity;
+import fr.free.nrw.commons.utils.FragmentUtils;
 import fr.free.nrw.commons.utils.ViewUtil;
 import io.reactivex.android.schedulers.AndroidSchedulers;
 import io.reactivex.disposables.Disposable;
@@ -108,17 +109,14 @@ public class SearchActivity extends NavigationBaseActivity implements MediaDetai
                     viewPager.setVisibility(View.VISIBLE);
                     tabLayout.setVisibility(View.VISIBLE);
                     searchHistoryContainer.setVisibility(View.GONE);
-                    if (searchImageFragment != null
-                        && searchImageFragment.isAdded()
-                        && !searchImageFragment.isDetached()) {
+                    if (FragmentUtils.isFragmentUIActive(searchImageFragment)) {
                         searchImageFragment.updateImageList(query.toString());
                     }
 
-                    if (searchCategoryFragment != null
-                        && searchCategoryFragment.isAdded()
-                        && !searchCategoryFragment.isDetached()) {
+                    if (FragmentUtils.isFragmentUIActive(searchCategoryFragment)) {
                         searchCategoryFragment.updateCategoryList(query.toString());
                     }
+
                 } else {
                     viewPager.setVisibility(View.GONE);
                     tabLayout.setVisibility(View.GONE);

--- a/app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java
@@ -12,16 +12,10 @@ import android.text.TextUtils;
 import android.view.View;
 import android.widget.FrameLayout;
 import android.widget.SearchView;
-
-import com.jakewharton.rxbinding2.view.RxView;
-import com.jakewharton.rxbinding2.widget.RxSearchView;
-
-import java.util.ArrayList;
-import java.util.List;
-import java.util.concurrent.TimeUnit;
-
 import butterknife.BindView;
 import butterknife.ButterKnife;
+import com.jakewharton.rxbinding2.view.RxView;
+import com.jakewharton.rxbinding2.widget.RxSearchView;
 import fr.free.nrw.commons.Media;
 import fr.free.nrw.commons.R;
 import fr.free.nrw.commons.explore.categories.SearchCategoryFragment;
@@ -31,6 +25,10 @@ import fr.free.nrw.commons.media.MediaDetailPagerFragment;
 import fr.free.nrw.commons.theme.NavigationBaseActivity;
 import fr.free.nrw.commons.utils.ViewUtil;
 import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.Disposable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
 
 /**
  * Represents search screen of this app
@@ -52,6 +50,7 @@ public class SearchActivity extends NavigationBaseActivity implements MediaDetai
     private MediaDetailPagerFragment mediaDetails;
     ViewPagerAdapter viewPagerAdapter;
     private String query;
+    private Disposable searchDisposable;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
@@ -98,28 +97,36 @@ public class SearchActivity extends NavigationBaseActivity implements MediaDetai
 
         viewPagerAdapter.setTabData(fragmentList, titleList);
         viewPagerAdapter.notifyDataSetChanged();
-        RxSearchView.queryTextChanges(searchView)
-                .takeUntil(RxView.detaches(searchView))
-                .debounce(500, TimeUnit.MILLISECONDS)
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe( query -> {
-                    this.query = query.toString();
-                            //update image list
-                            if (!TextUtils.isEmpty(query)) {
-                                viewPager.setVisibility(View.VISIBLE);
-                                tabLayout.setVisibility(View.VISIBLE);
-                                searchHistoryContainer.setVisibility(View.GONE);
-                                searchImageFragment.updateImageList(query.toString());
-                                searchCategoryFragment.updateCategoryList(query.toString());
-                            }else {
-                                viewPager.setVisibility(View.GONE);
-                                tabLayout.setVisibility(View.GONE);
-                                searchHistoryContainer.setVisibility(View.VISIBLE);
-                                recentSearchesFragment.updateRecentSearches();
-                                // open search history fragment
-                            }
-                        }
-                );
+        searchDisposable = RxSearchView.queryTextChanges(searchView)
+            .takeUntil(RxView.detaches(searchView))
+            .debounce(500, TimeUnit.MILLISECONDS)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(query -> {
+                this.query = query.toString();
+                //update image list
+                if (!TextUtils.isEmpty(query)) {
+                    viewPager.setVisibility(View.VISIBLE);
+                    tabLayout.setVisibility(View.VISIBLE);
+                    searchHistoryContainer.setVisibility(View.GONE);
+                    if (searchImageFragment != null
+                        && searchImageFragment.isAdded()
+                        && !searchImageFragment.isDetached()) {
+                        searchImageFragment.updateImageList(query.toString());
+                    }
+
+                    if (searchCategoryFragment != null
+                        && searchCategoryFragment.isAdded()
+                        && !searchCategoryFragment.isDetached()) {
+                        searchCategoryFragment.updateCategoryList(query.toString());
+                    }
+                } else {
+                    viewPager.setVisibility(View.GONE);
+                    tabLayout.setVisibility(View.GONE);
+                    searchHistoryContainer.setVisibility(View.VISIBLE);
+                    recentSearchesFragment.updateRecentSearches();
+                    // open search history fragment
+                }
+            });
     }
 
     /**
@@ -258,6 +265,14 @@ public class SearchActivity extends NavigationBaseActivity implements MediaDetai
     public void requestMoreImages() {
         if (searchImageFragment!=null){
             searchImageFragment.addImagesToList(query);
+        }
+    }
+
+    @Override protected void onDestroy() {
+        super.onDestroy();
+        //Dispose the disposables when the activity is destroyed
+        if (searchDisposable != null) {
+            searchDisposable.dispose();
         }
     }
 }

--- a/app/src/main/java/fr/free/nrw/commons/explore/images/SearchImageFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/images/SearchImageFragment.java
@@ -1,6 +1,5 @@
 package fr.free.nrw.commons.explore.images;
 
-
 import android.annotation.SuppressLint;
 import android.content.res.Configuration;
 import android.os.Bundle;
@@ -13,19 +12,9 @@ import android.view.View;
 import android.view.ViewGroup;
 import android.widget.ProgressBar;
 import android.widget.TextView;
-
-import com.pedrogomez.renderers.RVRendererAdapter;
-
-import java.util.ArrayList;
-import java.util.Date;
-import java.util.List;
-import java.util.concurrent.TimeUnit;
-
-import javax.inject.Inject;
-import javax.inject.Named;
-
 import butterknife.BindView;
 import butterknife.ButterKnife;
+import com.pedrogomez.renderers.RVRendererAdapter;
 import fr.free.nrw.commons.Media;
 import fr.free.nrw.commons.R;
 import fr.free.nrw.commons.di.CommonsDaggerSupportFragment;
@@ -37,7 +26,14 @@ import fr.free.nrw.commons.mwapi.OkHttpJsonApiClient;
 import fr.free.nrw.commons.utils.NetworkUtils;
 import fr.free.nrw.commons.utils.ViewUtil;
 import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.Disposable;
 import io.reactivex.schedulers.Schedulers;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Named;
 import timber.log.Timber;
 
 import static android.view.View.GONE;
@@ -66,6 +62,8 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
     OkHttpJsonApiClient okHttpJsonApiClient;
     @Inject @Named("default_preferences") BasicKvStore defaultKvStore;
 
+    private final String TAG="#SearchImageFragment#";
+
     private RVRendererAdapter<Media> imagesAdapter;
     private List<Media> queryList = new ArrayList<>();
 
@@ -75,6 +73,7 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
         ((SearchActivity)getContext()).onSearchImageClicked(index);
         saveQuery(query);
     });
+    private Disposable searchImagesDisposable;
 
     /**
      * This method saves Search Query in the Recent Searches Database.
@@ -140,11 +139,11 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
         bottomProgressBar.setVisibility(GONE);
         queryList.clear();
         imagesAdapter.clear();
-        okHttpJsonApiClient.searchImages(query, queryList.size())
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
-                .timeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
-                .subscribe(this::handleSuccess, this::handleError);
+        searchImagesDisposable = okHttpJsonApiClient.searchImages(query, queryList.size())
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .timeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
+            .subscribe(this::handleSuccess, this::handleError);
     }
 
 
@@ -230,8 +229,15 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
      * Handles the UI updates for no internet scenario
      */
     private void handleNoInternet() {
-        progressBar.setVisibility(GONE);
-        ViewUtil.showShortSnackbar(imagesRecyclerView, R.string.no_internet);
+        if (null
+            != getView()) {//We have exposed public methods to update our ui, we will have to add null checks until we make this lifecycle aware
+            if (null != progressBar) {
+                progressBar.setVisibility(GONE);
+            }
+            ViewUtil.showShortSnackbar(imagesRecyclerView, R.string.no_internet);
+        } else {
+            Timber.d(TAG + ": attempt to update fragment ui after its view was destroyed");
+        }
     }
 
     /**
@@ -257,6 +263,13 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
         }
         else {
             return imagesAdapter.getItem(i);
+        }
+    }
+
+    @Override public void onDestroyView() {
+        super.onDestroyView();
+        if(searchImagesDisposable!=null) {
+            searchImagesDisposable.dispose();//Always dispose the disposables when view is no longer there
         }
     }
 }

--- a/app/src/main/java/fr/free/nrw/commons/explore/images/SearchImageFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/images/SearchImageFragment.java
@@ -62,8 +62,6 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
     OkHttpJsonApiClient okHttpJsonApiClient;
     @Inject @Named("default_preferences") BasicKvStore defaultKvStore;
 
-    private final String TAG="#SearchImageFragment#";
-
     private RVRendererAdapter<Media> imagesAdapter;
     private List<Media> queryList = new ArrayList<>();
 
@@ -236,7 +234,7 @@ public class SearchImageFragment extends CommonsDaggerSupportFragment {
             }
             ViewUtil.showShortSnackbar(imagesRecyclerView, R.string.no_internet);
         } else {
-            Timber.d(TAG + ": attempt to update fragment ui after its view was destroyed");
+            Timber.d("Attempt to update fragment ui after its view was destroyed");
         }
     }
 

--- a/app/src/main/java/fr/free/nrw/commons/utils/FragmentUtils.java
+++ b/app/src/main/java/fr/free/nrw/commons/utils/FragmentUtils.java
@@ -15,6 +15,6 @@ public class FragmentUtils {
      * @return
      */
     public static boolean isFragmentUIActive(Fragment fragment) {
-        return fragment.getActivity() != null && fragment.isAdded() && !fragment.isDetached() && !fragment.isRemoving();
+        return fragment!=null && fragment.getActivity() != null && fragment.isAdded() && !fragment.isDetached() && !fragment.isRemoving();
     }
 }

